x-pack/legacy/plugins/lens/public/indexpattern_plugin/layerpanel.tsx

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import _ from 'lodash';
+import React, { useState, useMemo } from 'react';
+import {
+  EuiComboBox,
+  // @ts-ignore
+  EuiHighlight,
+  EuiButtonEmpty,
+  EuiIcon,
+  EuiIconTip,
+  EuiFlexGroup,
+  EuiFlexItem,
+} from '@elastic/eui';
+import { i18n } from '@kbn/i18n';
+import { I18nProvider } from '@kbn/i18n/react';
+import { DatasourceLayerPanelProps } from '../types';
+import { IndexPatternPrivateState } from './indexpattern';
+import { isLayerTransferable, updateLayerIndexPattern } from './state_helpers';
+
+export interface IndexPatternLayerPanelProps extends DatasourceLayerPanelProps {
+  state: IndexPatternPrivateState;
+  setState: (newState: IndexPatternPrivateState) => void;
+}
+
+export function LayerPanel({ state, setState, layerId }: IndexPatternLayerPanelProps) {
+  const [isChooserOpen, setChooserOpen] = useState(false);
+  const indexPatterns = useMemo(
```

Comment:
I don't think we need this useMemo. state.indexPatterns changes on each keypress when editing a field label, and on most interactions, as most interactions affect index pattern state.

I do wonder if we should flatten or normalize our state structure to reduce how much needless re-rendering occurs.

Anyway, in this case, we don't need to compute this at all unless the chooser is open. What I'd do instead is pull the entire chooser code out into its own component which can make this calculation internally. That way, it's only calculated while the chooser is open, and doesn't slow down rendering in the usual case.

Reply:
`state.indexPatterns` doesn't change on each keypress, this is just containing the available index patterns and fields. But you are right, it probably makes sense to only calculate this once the chooser is actually open. I will move that around.